contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,320 @@
+# Core Metrics Pipeline in kubelet
+
+**Author**: David Ashpole (@dashpole)
+
+**Last Updated**: 12/21/2016
+
+**Status**: Draft Proposal (WIP)
+
+This document proposes a design for an internal Core Metrics Pipeline.
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Core Metrics Pipeline in kubelet](#core-metrics-pipeline-in-kubelet)
+  - [Introduction](#introduction)
+    - [Background](#background)
+    - [Motivations](#motivations)
+    - [Proposal](#proposal)
+    - [Non Goals](#non-goals)
+  - [Design](#design)
+    - [Proposed Core Metrics API:](#proposed-core-metrics-api)
+    - [Core Machine Info:](#core-machine-info)
+  - [Implementation Plan](#implementation-plan)
+  - [Rollout Plan](#rollout-plan)
+  - [Implementation Status](#implementation-status)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Introduction
+
+### Background
+The [Monitoring Architecture](https://github.com/kubernetes/kubernetes/blob/master/docs/design/monitoring_architecture.md) proposal contains a blueprint for a set of metrics referred to as "Core Metrics".  The purpose of this proposal is to specify what those metrics are, and how they will be collected on the node.
+
+cAdvisor is an open source container monitoring solution which only monitors containers, and has no concept of k8s constructs like pods or volumes.  Kubernetes vendors cAdvisor into its codebase, and uses cAdvisor as a library with functions that enable it to collect metrics on containers.  The kubelet can then combine container-level metrics from cAdvisor with the kubelet's knowledge of k8s constructs like pods to produce the kubelet Summary statistics, which provides metrics for use by the kubelet, or by users through the [Summary API](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go).  cAdvisor works by collecting metrics at an interval (10 seconds, by default), and the kubelet then simply queries these cached metrics whenever it has a need for them.
+
+Currently, cAdvisor collects a large number of metrics related to system and container performance. However, only some of these metrics are consumed by the kubelet summary API, and many are not used.  The kubelet [Summary API](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go) is published to the kubelet summary API endpoint (stats/summary).  Some of the metrics provided by the summary API are consumed internally, but most are not used internally.
+
+### Motivations
```

Comment:
Another motivation is once we defined the new metrics API, we can denote the current Summary API for the monitoring pipeline. This means it can be iterated to provide more monitoring / introspection stats. There has been many requests to have a much richer set of monitoring stats from the node through Summary API, the reason we punt those enhancement is Summary API serves both purposes: system controlling and monitoring.